java/ddlogapi.c

```diff
@@ -13,6 +13,74 @@
 #include "ddlog.h"
 #include "ddlog_log.h"
 
+/* Error message returned by DDlog. */
+_Thread_local char* err_msg = NULL;
+
+/* 
+ * `vasprintf` implementation copied from
+ * https://github.com/littlstar/asprintf.c (MIT license)
+ */
+static int _vasprintf (char **str, const char *fmt, va_list args) {
+  int size = 0;
+  va_list tmpa;
+
+  va_copy(tmpa, args);
+  size = vsnprintf(NULL, 0, fmt, tmpa);
+  va_end(tmpa);
+
+  if (size < 0) { return -1; }
+
+  // alloc with size plus 1 for `\0'
+  *str = (char *) malloc(size + 1);
+
+  if (NULL == *str) { return -1; }
+
+  size = vsprintf(*str, fmt, args);
+  return size;
+}
+
+/* 
+ * Callback invoked by DDlog on error.
+ *
+ * Record error message in a thread-local variable so that it can later be
+ * wrapped in an exception.
+ */
+static void eprintln(const char* msg) {
+    if (err_msg) {
+        free(err_msg);
+    }
+    err_msg = strdup(msg);
+}
+
+/*
+ * Throw a `DDlogException` with error message in `err_msg`.
+ */
+static void throwDDlogException(JNIEnv* env) {
+    char * msg = err_msg ? err_msg : "Unknown error";
+    jclass eclass = (*env)->FindClass(env, "ddlogapi/DDlogException");
+    if (eclass) {
+        (*env)->ThrowNew(env, eclass, msg);
+    } else {
+        (*env)->ThrowNew(env, (*env)->FindClass(env, "java/lang/Exception"), msg);
+    }
+    err_msg = NULL;
+}
+
+static void throwIOException(JNIEnv* env, const char * fmt, ...) {
+    char *msg;
+    va_list ap;
+    va_start(ap, fmt);
+    int nbytes = _vasprintf(&msg, fmt, ap);
+    va_end(ap);
+
+    if (nbytes >= 0) {
+        (*env)->ThrowNew(env, (*env)->FindClass(env, "java/io/IOException"), msg);
+        free(msg);
+    } else {
+        (*env)->ThrowNew(env, (*env)->FindClass(env, "java/io/IOException"), "Unable to format error message");
```

Comment:
This should not happen, but if it does then the error will be very confusing.
Perhaps you want to say "exception occurred, but I could not show you the actual error message."

Reply:
good point; also the second one should have been `java.util.Exception`.  Thanks for spotting.